### PR TITLE
Introduce the first language page for the installer

### PR DESCRIPTION
**Pull request by greleuniforme**
The PR add the first page for the installer egg where you can choose the language.

### CODE REVIEW

File: locales/en/LC_MESSAGES/language_live_page.po [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+msgstr "Language:"
+
+msgid "desc_keyboard"
+msgstr "Keyboard:"
+
+msgid "input_keyboard"
+msgstr "Try the keyboard here"
+
+msgid "french"
+msgstr "french"
```
Comment (Arignir) [inline locales/en/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "French"
```

File: locales/en/LC_MESSAGES/language_live_page.po [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+msgid "english"
+msgstr "english"
```
Comment (Arignir) [inline locales/en/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "English"
```

File: locales/fr/LC_MESSAGES/language_live_page.po [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+"X-Poedit-SourceCharset: UTF-8\n"
+
+msgid "title"
+msgstr "Configuration de langue du CD autonomne"
+
+msgid "sidebar_title"
+msgstr "Langage CD autonomne"
+
+msgid "desc_language"
+msgstr "Langue:"
```
Comment (Arignir) [inline locales/fr/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "Langue :"
```

File: locales/fr/LC_MESSAGES/language_live_page.po [hunk 4]
```diff
@@ -0,0 +1,25 @@
+msgid ""
+msgstr ""
+"Content-Type: text/plain; charset=UTF-8\n"
+"X-Poedit-SourceCharset: UTF-8\n"
+
+msgid "title"
+msgstr "Configuration de langue du CD autonomne"
+
+msgid "sidebar_title"
+msgstr "Langage CD autonomne"
```
Comment (Arignir) [inline locales/fr/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "Langage de l'installeur"
```

File: locales/fr/LC_MESSAGES/language_live_page.po [hunk 5]
```diff
@@ -0,0 +1,25 @@
+msgid ""
+msgstr ""
+"Content-Type: text/plain; charset=UTF-8\n"
+"X-Poedit-SourceCharset: UTF-8\n"
+
+msgid "title"
+msgstr "Configuration de langue du CD autonomne"
```
Comment (Arignir) [inline locales/fr/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "Configuration du langage de l'installeur"
```

File: locales/fr/LC_MESSAGES/language_live_page.po [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+msgid "sidebar_title"
+msgstr "Langage CD autonomne"
+
+msgid "desc_language"
+msgstr "Langue:"
+
+msgid "desc_keyboard"
+msgstr "Clavier:"
```
Comment (Arignir) [inline locales/fr/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "Clavier :"
```

File: locales/fr/LC_MESSAGES/language_live_page.po [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+msgid "input_keyboard"
+msgstr "Essayez le clavier ici"
+
+msgid "french"
+msgstr "français"
```
Comment (Arignir) [inline locales/fr/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "Français"
```

File: locales/fr/LC_MESSAGES/language_live_page.po [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+msgid "english"
+msgstr "anglais"
```
Comment (Arignir) [inline locales/fr/LC_MESSAGES/language_live_page.po]:
```suggestion
msgstr "Anglais"
```

File: ui/gtk/main_window.py [hunk 9]
```diff
@@ -63,12 +82,134 @@ def __init__(self, locale_general: LanguageManagement, config_general: dict, con
...
         self._component.get_component('main_window').show_all()
+        GLib.idle_add(self.prepare_all_long_tasks_page)
 
     def set_title(self, message: str) -> None:
         self._component.get_component('main_window_header_right_label').set_halign(Gtk.Align.START)
         self._component.get_component('main_window_header_right_label').set_label(
             u'<span font-size=\'x-large\'>{}</span>'.format(message))
 
+    def register_all_pages(self) -> None:
+        all_pages = list()
```
Comment (Arignir) [inline ui/gtk/main_window.py]:
```suggestion
        all_pages = [
            LanguageLivePage(self._locale_general, self._config_general),
        ]
```

File: ui/gtk/main_window.py [hunk 10]
```diff
@@ -63,12 +82,134 @@ def __init__(self, locale_general: LanguageManagement, config_general: dict, con
...
+        if idx < 0:
+            return
+        self._page_index = idx
+        self.update_page()
+
+    def update_page(self) -> None:
+        self.set_title(self._pages[self._page_index].get_page_title())
+
+        if self._page_index == len(self._pages) - 1:
+            self.set_can_next(False)
```
Comment (Arignir) [inline ui/gtk/main_window.py]:
```suggestion
            self.set_can_next(self._page_index != len(self._pages) - 1)
```

(Same for all the `set_can_next`, `set_hide_next`, `set_can_prev`, `set_hide_prev` etc.)
(And obviously remove the outer condition)

File: ui/gtk/main_window.py [hunk 11]
```diff
@@ -63,12 +82,134 @@ def __init__(self, locale_general: LanguageManagement, config_general: dict, con
...
+        self._component.get_component('main_window_top_left_logo').set_from_pixbuf(pixbuf)
+        for label in self._component.get_component('main_window_left_list_text_box').get_children():
+            if label.get_page_id() == self._pages[self._page_index].get_page_id():
+                label.get_style_context().remove_class('dim-label')
+            else:
+                label.get_style_context().add_class('dim-label')
+        self._component.get_component('main_window_right_stack').set_visible_child_name(
+            self._pages[self._page_index].get_page_id())
+
+    def set_can_prev(self, can_prev: bool) -> None:
```
Comment (Arignir) [inline ui/gtk/main_window.py]: I'm not a fan of the `set_[verb]_[object]` pattern. It creates a million of function for all actions of all objects on the window.

An idea to help you out:
Create an enum `MainWindowButton` with all the buttons, for example `NEXT='main_window_next_btn'`, etc. and create functions like `set_button_visibility(button: MainWindowButton, status: Bool)`
Do not add the condition `if self.page_index` as you're already doing that condition when calling the function

Note that the term `set_hide_xxx` while taking a parameter that states if we want to hide or show the element seems conflictual: the parameters don't match the function name. Prefer neutral names like 'set_visibility', etc.

File: ui/gtk/main_window.py [hunk 12]
```diff
@@ -63,12 +82,134 @@ def __init__(self, locale_general: LanguageManagement, config_general: dict, con
...
+            self._pages.append(current_page)
+
+    def load_lang(self) -> None:
+        component = self._component.get_component('main_window_left_list_text_box')
+        for label in component.get_children():
+            component.remove(label)
+
+        for current_page in self._pages:
+            label_in_box = TitleLabel(current_page.get_page_id(), current_page.get_page_sidebar_title())
+            if label_in_box.get_page_id() == self._pages[self._page_index].get_page_id():
```
Comment (Arignir) [inline ui/gtk/main_window.py]: Please comment non-obvious parts of your code. What is this doing? Why?

**Review by Arignir — COMMENTED** [review]
Good job!

Try to put more efforts on the translations, missing the capital letters is a Yikes that can easily be avoided.

You're using the `self._pages[self._page_index]` pattern a shit ton: create a function! You can name it `current_page()` or stg like that.

Also, I think you forgot to push the `LanguageLivePage` class (I couldn't find it, at least), and it's parent class if it has one (I suspect it does).

File: locales/en/LC_MESSAGES/language_live_page.po [hunk 13]
```diff
@@ -10,16 +10,16 @@ msgid "sidebar_title"
 msgstr "Live CD language"
 
 msgid "desc_language"
-msgstr "Language:"
+msgstr "Language :"
```
Comment (Arignir) [inline locales/en/LC_MESSAGES/language_live_page.po:13]: Nope, in English there's no space before the punctuation like `:`, `!`, `?`,  etc.

There is one in French, though!

File: ui/gtk/main_window.py [hunk 14]
```diff
@@ -147,55 +154,34 @@ def prev_page_btn(self) -> None:
...
-            self.set_hide_prev(True)
-        else:
-            self.set_can_prev(True)
-            self.set_hide_prev(False)
-
-        self._pages[self._page_index].load_page()
-
-        pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_scale(self._pages[self._page_index].get_page_icon(), 50, 50, False)
+        self.set_title(self.current_page().get_page_title())
+        self.set_button_action_visibility(MainWindowButton.NEXT, self._page_index != len(self._pages) - 1)
```
Comment (Arignir) [inline ui/gtk/main_window.py:158]: Excellent!

File: ui/gtk/main_window_button.py [hunk 15]
```diff
@@ -0,0 +1,6 @@
+from enum import Enum
+
+
+class MainWindowButton(Enum):
```
Comment (Arignir) [inline ui/gtk/main_window_button.py:4]: You can probably move this enum to `main_window.py`, that would eliminate this file which doesn't add much, and keep everything in a single module.

**Review by Arignir — CHANGES_REQUESTED** [review]

**Comment by Arignir** [comment]
You have taken the suggestions seriously and there's some really good changes here. Keep up the good work!

File: ui/gtk/main_window_button.py (hunk 15 above)
Comment (greleuniforme) [reply ui/gtk/main_window_button.py:4]: Can't because the pages will include main_window and main_window will include pages.

**Comment by Arignir** [comment]
I don't know if you or I fucked up something, but the last comment you pushed (63e0bc8) is actually empty.

**Comment by Arignir** [comment]
(I still see nothing :o Can you confirm it's empty on your side?)

**Pull request closed without merging**